File: uniquePaths.py
# problem: https://leetcode.com/problems/unique-paths/
# Runtime: 32 ms, faster than 69.60% of Python3 online submissions for Unique Paths.
# Memory Usage: 14.1 MB, less than 96.34% of Python3 online submissions for Unique Paths.

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        dp = [[0 for _ in range(n)] for _ in range(m)]
        dp[0][0] = 1
        for i in range(m):
            for j in range(n):
                if i > 0:
                    dp[i][j] += dp[i-1][j]
                if j > 0:
                    dp[i][j] += dp[i][j-1]
        return dp[-1][-1]

File: uniquePaths.cpp
#include <bits/stdc++.h>

using namespace std;


class Solution {
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[0][0] = 1;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                int up = i > 0 ? dp[i - 1][j] : 0;
                int left = j > 0 ? dp[i][j - 1] : 0;
                dp[i][j] += up + left;
            }
        }
        return dp[m - 1][n - 1];
    }
};

